Better handling of alias analysis for aten::backward

smessmer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #36683 Use overload names for aten::backward
* **#36522 Better handling of alias analysis for aten::backward**

Differential Revision: [D21004102](https://our.internmc.facebook.com/intern/diff/D21004102/)

smessmer committed:
Better handling of alias analysis for aten::backward



Differential Revision: [D21004102](https://our.internmc.facebook.com/intern/diff/D21004102/)

[ghstack-poisoned]

smessmer committed:
Update on "Better handling of alias analysis for aten::backward"


Differential Revision: [D21004102](https://our.internmc.facebook.com/intern/diff/D21004102/)

[ghstack-poisoned]

bhosmer commented on torch/csrc/autograd/VariableTypeManual.cpp:
```diff
@@ -375,7 +375,7 @@ static auto registry = torch::RegisterOperators()
   .op(torch::RegisterOperators::options()
     .schema("aten::backward(Tensor self, Tensor? gradient=None, bool keep_graph=False, bool create_graph=False) -> ()")
-    .aliasAnalysis(AliasAnalysisKind::FROM_SCHEMA)
+    .aliasAnalysis(AliasAnalysisKind::CONSERVATIVE)
```
Hmmm, I was expecting to see `Tensor(a!) self` in the schema. As given, won't the change to `CONSERVATIVE` change behavior?

smessmer replied:
CONSERVATIVE implies that every argument is mutable and aliased with every other argument. The `(a!)` was actually unnecessary.

smessmer replied:
ah that's still part of the misexport. There is a location that did specify the `(a!)`. I now added its removal to the PR.

bhosmer replied:
Can we not use the same string literal for this registration and equivalent (second) one in -[register_prim_ops_fulljit](https://github.com/pytorch/pytorch/pull/36522/files#diff-fb26a0de03180fc8778ce18261715d98R153)? Even now they're not quite the same :)

smessmer replied:
fixed

bhosmer commented on torch/csrc/jit/runtime/operator.h:
```diff
@@ -149,20 +149,12 @@ struct TORCH_API Operator {
-          schemaRef,
-          " with aliasing information in the schema but without AliasAnalysisKind::FROM_SCHEMA.");
-    }
+    TORCH_CHECK(
```
Similarly, I'm not sure why the assert dodge was necessary at all, given that the schema for backward didn't have aliasing info (unless I'm misreading)

albanD replied:
I think the (a!) was aliasing info no?

wanchaol replied:
yeah but this PR deleted `(a!)` from the schema and use conservative aliasing, so backward is not having any aliasing information i think.

wanchaol commented:
Should we also remove the special casing on `aten::backward` in `aliasAnalysisKind()`?

smessmer committed:
Update on "Better handling of alias analysis for aten::backward"


Differential Revision: [D21004102](https://our.internmc.facebook.com/intern/diff/D21004102/)

[ghstack-poisoned]

smessmer commented:
@wanchaol that's what this PR is doing right? Or do you mean some other `aliasAnalysisKind`?

smessmer committed:
Update on "Better handling of alias analysis for aten::backward"


Differential Revision: [D21004102](https://our.internmc.facebook.com/intern/diff/D21004102/)

[ghstack-poisoned]

smessmer committed:
Update on "Better handling of alias analysis for aten::backward"


Differential Revision: [D21004102](https://our.internmc.facebook.com/intern/diff/D21004102/)

[ghstack-poisoned]

wanchaol commented on torch/csrc/jit/runtime/register_prim_ops_fulljit.cpp:
```diff
@@ -150,7 +150,7 @@ RegisterOperators reg(
          aliasAnalysisConservative()),
      Operator(
-         "aten::backward(Tensor(a!) self, Tensor? gradient=None, bool? retain_graph=None, bool create_graph=False) -> ()",
+         "aten::backward(Tensor self, Tensor? gradient=None, bool? retain_graph=None, bool create_graph=False) -> ()",
```
I guess we could try to delete this overload here, as in VariableTypeManual we already this as an operator.

wanchaol reviewed (commented):
> @wanchaol that's what this PR is doing right? Or do you mean some other `aliasAnalysisKind`?

you are right, sorry I didn't pay close attention to it last time. This looks good to me, I am wondering if we could delete the jit operator as we registered the operator in VariableTypeManual. 

Edit: Feature request on operator registration, I think it's better for operator registration to have this functionality: If I registered the same operator twice (with the same operator name and schema), we should just throw error/warning. It's good to warn the user that we already have the same operator otherwise overload resolution might be confusing.

smessmer commented:
superseded by superseded by https://github.com/pytorch/pytorch/pull/36750

smessmer commented:
> Edit: Feature request on operator registration, I think it's better for operator registration to have this functionality: If I registered the same operator twice (with the same operator name and schema), we should just throw error/warning. It's good to warn the user that we already have the same operator otherwise overload resolution might be confusing.

agreed, I'm working on that on the side, but there's currently a lot of ops that do register themselves multiple times... :(
